File: jar/src/main/java/jline/api/cache/Cache_gamma_lp.java
/**
 * @file Cache Access Factor Computation via Linear Programming
 *
 * @since LINE 3.0
 */
package jline.api.cache;

import java.util.ArrayList;

import jline.io.Ret;
import jline.util.matrix.Matrix;

public final class Cache_gamma_lp {
    private Cache_gamma_lp() {}

    /**
     * Computes access factors for the cache.
     */
    public static Ret.cacheGamma cache_gamma_lp(Matrix[] lambda, Matrix[][] R) {
        int u = lambda.length;
        int n = lambda[0].getNumRows();
        int h = lambda[0].getNumCols() - 1;
        Matrix gamma = new Matrix(n, h);

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < h; j++) {
                Matrix Rvi = new Matrix(R[0][i].getNumRows(), R[0][i].getNumCols());
                for (int v = 0; v < u; v++) {
                    Rvi = Rvi.add(1.0, R[v][i]);
                }
                ArrayList<Integer> Pij = new ArrayList<Integer>();
                Pij.add(1 + j);
                ArrayList<Integer> pr_j = cache_par(Rvi, 1 + j);
                while (pr_j.size() > 0) {
                    Pij.add(0, pr_j.get(0));
                    pr_j = cache_par(Rvi, pr_j.get(0));
                }
                if (Pij.size() == 0) {
                    gamma.set(i, j, 0);
                } else {
                    gamma.set(i, j, 1);
                    for (int li = 1; li < Pij.size(); li++) {
                        double y = 0.0;
                        int l_1 = Pij.get(li - 1);
                        int l = Pij.get(li);
                        for (int v = 0; v < u; v++) {
                            for (int t = 0; t <= l_1; t++) {
                                y += lambda[v].get(i, t) * R[v][i].get(t, l);
                            }
                        }
                        gamma.set(i, j, gamma.get(i, j) * y);
                    }
                }
            }
        }
        int[] parent = new int[h];
        // tree structure read off item 0's routing matrix aggregated over users,
        // the same matrix the gamma loop walks
        Matrix Rtot = new Matrix(R[0][0].getNumRows(), R[0][0].getNumCols());
        for (int v = 0; v < u; v++) {
            Rtot = Rtot.add(1.0, R[v][0]);
        }
        for (int j = 0; j < h; j++) {
            ArrayList<Integer> pj = cache_par(Rtot, 1 + j);
            parent[j] = pj.isEmpty() ? -1 : (pj.get(0) - 1); // list indices, -1 = miss list
        }
        return new Ret.cacheGamma(gamma, u, n, h, parent);
    }

    /**
     * Finds the parent of a given list index.
     */
    public static ArrayList<Integer> cache_par(Matrix R, int j) {
        ArrayList<Integer> parent = new ArrayList<Integer>();
        for (int i = 0; i <= j - 1; i++) {
            if (R.get(i, j) != 0.0) {
                parent.add(i);
            }
        }
        if (parent.size() > 1) {
            throw new RuntimeException("A cache has a list with more than one parent, but the structure must be a tree.");
        }
        return parent;
    }
}

File: jar/src/main/kotlin/jline/api/cache/Cache_gamma_lp.kt
/**
 * @file Cache Access Factor Computation via Linear Programming
 * 
 * Computes cache access factors using linear programming optimization methods.
 * Access factors are essential parameters that characterize request patterns
 * and routing probabilities in multi-level cache systems.
 * 
 * @since LINE 3.0
 */
package jline.api.cache

import jline.io.Ret
import jline.util.matrix.Matrix

/**
 * Computes access factors for the cache.
 *
 * @param lambda - Matrix array representing request arrival rates from users to items of individual lists.
 * @param R      - 2D Matrix array representing the reachability graph of a list for different streams and items.
 * @return cacheGammaLpReturn - An object containing access factors (gamma), the number of users (u),
 * the number of items (n), and the number of lists (h).
 */

fun cache_gamma_lp(lambda: Array<Matrix>, R: Array<Array<Matrix>>): Ret.cacheGamma {
    val u = lambda.size // Number of users
    val n = lambda[0].numRows // Number of items
    val h = lambda[0].numCols - 1 // Number of lists
    val gamma = Matrix(n, h)

    for (i in 0..<n) { // for all items
        for (j in 0..<h) { // for all levels
            // Compute gamma(i,j)
            var Rvi = Matrix(R[0][i].numRows, R[0][i].numCols)
            for (v in 0..<u) {
                Rvi = Rvi.add(1.0, R[v][i])
            }
            val Pij = ArrayList<Int>()
            Pij.add(1 + j)
            var pr_j = cache_par(Rvi, 1 + j)
            while (pr_j.size > 0) {
                Pij.add(0, pr_j[0])
                pr_j = cache_par(Rvi, pr_j[0])
            }
            if (Pij.size == 0) {
                gamma[i, j] = 0
            } else {
                gamma[i, j] = 1
                for (li in 1..<Pij.size) { // For all levels up to the current one
                    var y = 0.0
                    val l_1 = Pij[li - 1]
                    val l = Pij[li]
                    for (v in 0..<u) { // For all streams
                        for (t in 0..l_1) {
                            y += lambda[v][i, t] * R[v][i][t, l]
                        }
                    }
                    gamma[i, j] = gamma[i, j] * y
                }
            }
        }
    }
    return Ret.cacheGamma(gamma, u, n, h)
}

/**
 * Finds the parent of a given list index according to the access probabilities in an access probability matrix.
 *
 * @param R - Matrix representing the access probabilities.
 * @param j - Integer representing the list for which the parent will be found.
 * @return ArrayList<Integer> - A list containing the parent of the specified list index j.
 * @throws RuntimeException if a list has more than one parent.
</Integer> */
fun cache_par(R: Matrix, j: Int): ArrayList<Int> {
    val parent = ArrayList<Int>()
    for (i in 0..j - 1) {
        if (R[i, j] != 0.0) {
            parent.add(i)
        }
    }
    if (parent.size > 1) {
        throw RuntimeException("A cache has a list with more than one parent, but the structure must be a tree.")
    }
    return parent
}
/**
 * Cache gamma lp algorithms
 */
@Suppress("unused")
class CacheGammaLpAlgo {
    companion object {
        // Class documentation marker for Dokka
    }
}